diffrax/_global_interpolation.py

```diff
@@ -134,13 +135,22 @@ def evaluate(
         def _index(_ys):
             return _ys[index]
 
+        y_leaves = jtu.tree_leaves(self.ys)
+        if len(y_leaves) == 0:
+            ys_dtype = default_floating_dtype()
+        else:
+            ys_dtype = jnp.result_type(*y_leaves)
```

Comment:
I think this is wrong: this will result in the entire operation below getting cast to the same largest dtype.
Rather, the casting should happen on a per-leaf basis, so that we do the minimal amount of upcasting necessary.

(If this is just to avoid the strict-dtype-casting rules, then I'm okay with locally disabling that using `with jax....` as discussed elsewhere.)

Reply:
(Actually, I'm realising that means we probably don't need `default_floating_dtype` at all, as no-leaves means that we have a no-op.)